pynamodb/connection/base.py

```diff
@@ -465,7 +466,20 @@ def _make_api_call(self, operation_name: str, operation_kwargs: Dict, settings:
                     verbose_properties['table_name'] = operation_kwargs.get(TABLE_NAME)
 
                 try:
-                    raise VerboseClientError(botocore_expected_format, operation_name, verbose_properties)
+                    raise VerboseClientError(
+                        botocore_expected_format,
+                        operation_name,
+                        verbose_properties,
+                        cancellation_reasons=(
+                            (
+                                CancellationReason(
+                                    code=d['Code'],
+                                    message=d.get('Message'),
+                                ) if d['Code'] != 'None' else None
```

Comment:
Why return `None` here? Shouldn't we instead not return anything at all?

The following code will break with `AttributeError: 'NoneType' object has no attribute 'code'`:
```
        except pydb_exc.TransactWriteError as e:
            retry_err_types = (
                'ProvisionedThroughputExceededException',
                'ThrottlingError',
                'TransactionConflict',
            )
            for error in e.cancellation_reasons:
                if error.code in retry_err_types:
                    raise RuntimeError(f'transaction failed {error.code}')
```
Just seems like it'd be cleaner to not return anything vs returning `None`

Reply:
The cancellation reasons are 1:1 to the transaction items so you can know what item was the reason.

Would you rather the reason itself not to be none but for code to be none?

Reply:
Yeah to me that would feel cleaner but I guess it's subjective. Just negates the extra "if error and error.code"

Reply:
The thing is, if it was "cleaner", it would leave  users with no way of associating reasons with individual transaction items. Maybe a different interface would convey this better, e.g. mapping of item to reason.

Reply:
yeah that would make sense to me